pins/boards.py

```diff
@@ -326,7 +310,55 @@ def pin_write(
 
         return meta
 
-    def pin_download(self, name, version=None, hash=None):
+    def pin_write(
+        self,
+        x,
+        name: Optional[str] = None,
+        type: Optional[str] = None,
+        title: Optional[str] = None,
+        description: Optional[str] = None,
+        metadata: Optional[Mapping] = None,
+        versioned: Optional[bool] = None,
+        created: Optional[datetime] = None,
+    ) -> Meta:
+        """Write a pin object to the board.
+
+        Parameters
+        ----------
+        x:
+            An object (e.g. a pandas DataFrame) to pin.
+        name:
+            Pin name.
+        type:
+            File type used to save `x` to disk. May be "csv", "arrow", "parquet",
+            "joblib", "json", or "file".
+        title:
+            A title for the pin; most important for shared boards so that others
+            can understand what the pin contains. If omitted, a brief description
+            of the contents will be automatically generated.
+        description:
+            A detailed description of the pin contents.
+        metadata:
+            A dictionary containing additional metadata to store with the pin.
+            This gets stored on the Meta.user field.
+        versioned:
+            Whether the pin should be versioned. Defaults to versioning.
+        created:
+            A date to store in the Meta.created field. This field may be used as
+            part of the pin version name.
+        """
+
+        if type == "file":
+            raise NotImplementedError(
+                ".pin_write() does not support type='file'. "
+                "Use .pin_upload() to save a file as a pin."
+            )
+
+        return self._pin_store(
+            x, name, type, title, description, metadata, versioned, created
+        )
+
+    def pin_download(self, name, version=None, hash=None) -> Sequence[str]:
```

Comment:
This is mostly a question for my understanding: is the outcome of `pin_download` to help find the path _to_ download the pin, rather than actually downloading the pin?

Reply:
It downloads the pin, and returns the path to the downloaded pin.

There's a sort of important architecture piece here. In R pins, content always has to be downloaded first and put onto disk somewhere. Python pins uses a file object, which hides the implementation behavior of how the content is retrieved. e.g. `some_file.read()` can save to an on-disk cache if it wants to--and that's the default behavior--but it's not necessary. However, because `pin_download()` requires returning a path to a file on disk, python pins can only do that in cases where there is an on-disk cache.

That's why we need this lil cache check:

```python
        # could also check whether f isinstance of PinCache
        fname = getattr(f, "name", None)

        if fname is None:
            raise PinsError("pin_download requires a cache.")
```

Reply:
Ah, yes! Thanks for the explanation. That all makes sense--I suppose I was looking for a user-defined path where things got downloaded to, rather than getting downloaded to a set location (cache).